xtra-libp2p/tests/load.rs

```diff
@@ -0,0 +1,392 @@
+            done_times_count: 0,
+        }
+    }
+}
+
+#[xtra_productivity(message_impl = false)]
+impl SomeMessageExchangeListener {
+    async fn handle(&mut self, msg: NewInboundSubstream, ctx: &mut Context<Self>) {
+        tracing::info!("{} handling NewInboundSubstream", self.protocol);
+
+        let this = ctx.address().expect("self to be alive");
+
+        let protocol = self.protocol.clone();
+        let future = {
+            let this = this.clone();
+            async move {
+                some_message_exchange_listener(msg.stream).await?;
+                this.send_async_safe(ListenerDone).await?;
+
+                tracing::info!("{} sent ListenerDone", protocol);
+
+                anyhow::Ok(())
+            }
+        };
+
+        tokio_extras::spawn_fallible(&this, future, move |e| async move {
+            tracing::warn!("Parallel message with peer {} failed: {}", msg.peer, e);
+        });
+    }
+}
+
+#[xtra_productivity]
+impl SomeMessageExchangeListener {
+    async fn handle(&mut self, _: ListenerDone) {
+        self.done_times_count += 1;
+
+        tracing::info!(
+            "{} received ListenerDone for trigger time {}",
+            self.protocol,
+            self.done_times_count
+        );
+    }
+
+    async fn handle(&mut self, check_done: CheckListenerDone) -> bool {
+        self.done_times_count == check_done.0
+    }
+}
+
+#[async_trait]
+impl Actor for SomeMessageExchangeListener {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+struct ListenerDone;
+struct CheckListenerDone(u64);
+
+pub fn into_arr<T, const N: usize>(v: Vec<T>) -> [T; N] {
+    v.try_into()
+        .unwrap_or_else(|v: Vec<T>| panic!("Expected a Vec of length {} but it was {}", N, v.len()))
+}
+
+async fn some_message_exchange_dialer(stream: xtra_libp2p::Substream) -> Result<()> {
+    let mut stream =
+        asynchronous_codec::Framed::new(stream, asynchronous_codec::LengthCodec).fuse();
+
+    stream.send(Bytes::from("Dialer")).await?;
+    let bytes = stream
+        .select_next_some()
+        .await
+        .context("Expected message")?;
+    let message = String::from_utf8(bytes.to_vec())?;
+
+    assert_eq!(message, "Listener");
+
+    Ok(())
+}
+
+async fn some_message_exchange_listener(stream: xtra_libp2p::Substream) -> Result<()> {
+    let mut stream =
+        asynchronous_codec::Framed::new(stream, asynchronous_codec::LengthCodec).fuse();
+
+    let bytes = stream.select_next_some().await?;
+    let name = String::from_utf8(bytes.to_vec())?;
+
+    stream.send(Bytes::from("Listener")).await?;
+
+    assert_eq!(name, "Dialer");
+
+    Ok(())
+}
+
+pub fn init_tracing() -> DefaultGuard {
+    tracing_subscriber::fmt()
+        .with_env_filter("WARN")
+        .with_env_filter("load=debug")
+        .with_env_filter("xtra=debug")
+        .with_env_filter("xtra_libp2p=debug")
+        .with_env_filter("xtras=debug")
+        .with_test_writer()
+        .set_default()
```

Comment:
I purposely created this PR without the Jaeger instrumentation because Cael was still working on it on his fork which makes it a bit hard to depend on without creating conflicts.
But I think these tests are actually independent of the other instrumentation, i.e. these tests have a separate instrumentation setup, so I can bring that commit back into this PR :)
Sorry, did not think that through enough, I will see to add the commit that adds the Jaeger instrumentation.

Reply:
After talking to @Restioson I came to the conclusion that which ever PR comes first, this one or https://github.com/itchysats/itchysats/pull/2404 will require some rebase in the other. 

> But I think these tests are actually independent of the other instrumentation

Counter to my comment above we should actually make use of the macro that is added in #2404 so I don't intend to add the instrumentation in this PR unless #2404 merges first :)